src/main/cpp/src/number_converter.cu

```diff
@@ -0,0 +1,761 @@
+    // const bases, check bases only once
+    if (is_invalid_base_range(from_base.get(0), to_base.get(0))) {
+      // if base is invalid, return all nulls thus no overflow
+      return false;
+    }
+  }
+
+  return thrust::any_of(
+    rmm::exec_policy_nosync(stream),
+    thrust::counting_iterator<cudf::size_type>(0),
+    thrust::counting_iterator<cudf::size_type>(num_rows),
+    is_overflow_fn<STR_ITERATOR, FROM_BASE_ITERATOR, TO_BASE_ITERATOR, IS_CONST_BASES>{
+      input, from_base, to_base});
+}
+
+bool is_cv(convert_number_t const& t) { return std::holds_alternative<cudf::column_view>(t); }
+
+void check_types(convert_number_t const& input,
+                 convert_number_t const& from_base,
+                 convert_number_t const& to_base,
+                 rmm::cuda_stream_view stream)
+{
+  // check input type
+  if (is_cv(input)) {
+    auto const input_cv = std::get<cudf::column_view>(input);
+    CUDF_EXPECTS(input_cv.type().id() == cudf::type_id::STRING,
+                 "Input column must be of type STRING");
+  } else {
+    auto const input_scalar = std::get<cudf::string_scalar>(input);
+    CUDF_EXPECTS(input_scalar.type().id() == cudf::type_id::STRING,
+                 "Input scalar must be of type STRING");
+    CUDF_EXPECTS(input_scalar.is_valid(stream), "Input scalar must be valid");
+  }
+
+  // check from_base type
+  if (is_cv(from_base)) {
+    auto const from_base_cv = std::get<cudf::column_view>(from_base);
+    CUDF_EXPECTS(from_base_cv.type().id() == cudf::type_id::INT32,
+                 "From base column must be of type INT32");
+  }
+
+  // check to_base type
+  if (is_cv(to_base)) {
+    auto const to_base_cv = std::get<cudf::column_view>(to_base);
+    CUDF_EXPECTS(to_base_cv.type().id() == cudf::type_id::INT32,
+                 "To base column must be of type INT32");
+  }
+}
+
+}  // anonymous namespace
+
+std::unique_ptr<cudf::column> convert(convert_number_t const& input,
+                                      convert_number_t const& from_base,
+                                      convert_number_t const& to_base,
+                                      rmm::cuda_stream_view stream,
+                                      rmm::device_async_resource_ref mr)
+{
+  check_types(input, from_base, to_base, stream);
+
+  if (is_cv(input)) {
+    auto const input_cv = std::get<cudf::column_view>(input);
+    auto const d_strs   = cudf::column_device_view::create(input_cv, stream);
+
+    if (is_cv(from_base)) {
+      auto const from_base_cv = std::get<cudf::column_view>(from_base);
+      auto const d_from_bases = cudf::column_device_view::create(from_base_cv, stream);
+
+      if (is_cv(to_base)) {
+        auto const to_base_cv = std::get<cudf::column_view>(to_base);
+        auto const d_to_bases = cudf::column_device_view::create(to_base_cv, stream);
+        // input is string cv, from base is cv, to base is cv
+        return convert_impl<str_iter, base_iter, base_iter, /*IS_CONST_BASES*/ false>(
```

Comment:
The template parameters can be omitted as the compiler can infer them.
```suggestion
        return convert_impl(
```

Reply:
Similar for all other places.

Reply:
Done